romana/cmd/root.go

```diff
@@ -0,0 +1,184 @@
+designed for Cloud Native applications. Romana allows multi-tenant
+cloud computing networks for OpenStack, Docker and Kubernetes to
+be built without encapsulation or a virtual network overlay.
+
+Romana networks are less expensive to build, easier to operate
+and deliver higher performance than networks built using
+alternative overlay based SDN designs.
+
+For more information, please check http://romana.io
+`,
+}
+
+// Execute adds all child commands to the root command sets flags appropriately.
+// This is called by main.main(). It only needs to happen once to the rootCmd.
+func Execute() {
+	if err := RootCmd.Execute(); err != nil {
+		fmt.Println(err)
+		os.Exit(-1)
+	}
+}
+
+func init() {
+	cobra.OnInitialize(initConfig)
+
+	RootCmd.AddCommand(hostCmd)
+	RootCmd.AddCommand(tenantCmd)
+	RootCmd.AddCommand(segmentCmd)
+
+	RootCmd.Flags().BoolVarP(&version, "version", "v",
+		false, "Build and Versioning Information.")
+
+	RootCmd.PersistentFlags().StringVarP(&cfgFile, "config",
+		"c", "", "config file (default is $HOME/.romana.yaml)")
+	RootCmd.PersistentFlags().StringVarP(&rootURL, "rootURL",
+		"r", "", "root service url, e.g. http://192.168.0.1")
+	RootCmd.PersistentFlags().StringVarP(&format, "format",
+		"f", "", "enable formatting options like [json|table], etc.")
+	RootCmd.PersistentFlags().StringVarP(&platform, "platform",
+		"p", "", "Use platforms like [openstack|kubernetes], etc.")
+	RootCmd.PersistentFlags().BoolVarP(&verbose, "verbose",
+		"w", false, "Verbose output.")
+
+	RootCmd.PersistentPreRun = preConfig
+	RootCmd.Run = versionInfo
+}
+
+// preConfig sanitizes URLs and setup viper with URLs.
+func preConfig(cmd *cobra.Command, args []string) {
+	// Add port details to rootURL else try localhost
+	// if nothing is given on command line or config.
+	if rootURL == "" {
+		rootURL = viper.GetString("RootURL")
+	}
+	if rootURL != "" {
+		rootURL = strings.TrimSuffix(rootURL, "/")
+		rootURL = rootURL + ":9600/"
+	} else {
+		rootURL = "http://localhost:9600/"
+	}
+	viper.Set("RootURL", rootURL)
+
+	// Give command line options higher priority then
+	// the corresponding config options.
+	if format == "" {
+		format = viper.GetString("Format")
+	}
+	// if format is still not found just default to tabular format.
+	if format == "" {
+		format = "table"
+	}
+	viper.Set("Format", format)
+
+	if platform == "" {
+		platform = viper.GetString("Platform")
+	}
+	if platform == "" {
+		platform = "openstack"
+	}
+	viper.Set("Platform", platform)
+
+}
+
+// versionInfo displays the build and versioning information.
+func versionInfo(cmd *cobra.Command, args []string) {
+	if version {
+		fmt.Println(common.BuildInfo())
+		os.Exit(0)
+	}
+	cmd.Help()
+}
+
+// initConfig reads in config file and ENV variables if set.
+func initConfig() {
+	if cfgFile != "" { // enable ability to specify config file via flag
+		viper.SetConfigFile(cfgFile)
+	}
+
+	viper.SetConfigName(".romana") // name of config file (without extension)
```

Comment:
We want to use ".romana" as extension? Isn't it usually the other way around, such as "romana.conf"?